### PR TITLE
Draft of Keras pickle RFC

### PR DESCRIPTION
**Pull request by stsievert**
**What does this PR implement?**
A draft of the Keras pickle edits. I'm submitting this PR because I want to discuss these changes and don't want add a bunch of unnecessary comments/notifications/reviews that the TF maintainers have to sift through.

This PR should be merged when the first draft is done.

**Reference issues/PRs**
This is a draft for https://github.com/tensorflow/community/pull/286

### CODE REVIEW

**Comment by adriangb** [comment]
I'm sorry that I just saw this and in the meantime made edits to the original. My apologies. I will try to incorporate your version into mine.

**Comment by adriangb** [comment]
Sorry for rewriting the history @stsievert . I incorporated what you wrote into what I had. We can continue working here until we have something more final that we can upstream.

**Comment by adriangb** [comment]
I think I'd like to also include a monkey patched example for using `Model.save`, using a string to transfer back and forth, just as a POC.

**Comment by adriangb** [comment]
POC for using SaveModel as a backend for pickle is up in the notebook: https://colab.research.google.com/drive/1SEGXDFfNl5i0Cy8a4xvRjWOOjdq3sOOC?authuser=1#scrollTo=VZIkAZyJE1CP

**Comment by adriangb** [comment]
Anything else you think we need to add @stsievert?

**Comment by stsievert** [comment]
I think so. I’d like to make some edits to better frame the problem. I’d also like a chance to review the technical content.

**Comment by adriangb** [comment]
Okay, I'll let you do the next round of edits before I make any more changes.

**Comment by stsievert** [comment]
I made some edits to the motivation and user benefits section. I'd still like to review the technical content.

**Comment by adriangb** [comment]
Edits look great. And thanks for fixing the line lengths.

**Comment by stsievert** [comment]
Why implement `__reduce_ex__`? `__reduce_ex__` only exists to support different Pickle versions ([docs][1]). AFAICK the argument `protocol` isn't used. Why not `__reduce__`?

I'd still like to make a couple more edits to this RFC. I'll try to make the edits later tomorrow. Feel free to make some edits; I revised the technical section.

[1]:https://docs.python.org/3/library/pickle.html#object.__reduce_ex__

**Comment by adriangb** [comment]
Two main reasons:
1. I recall running into issues using just `__reduce__` when I first tried that. I think it was breaking some test that implemented it in subclassed models, but I do not remember tbh. Might be worth trying again.
2. I think it's likely that they will want to use that protocol version in the future, ex: to map it to different versions of the SaveModel protocol.

**Comment by stsievert** [comment]
@mrocklin we're proposing to add Pickle support to Tensorflow. Could review our proposal? It's at [20200902-pickle-for-keras.md][1]. I'd appreciate any of your thoughts, including any ideas on other problems Pickle support would solve for Keras.

In this, I lifted some wording about ecosystem support from your post "[Pickle isn't slow, it's a protocol]." Here's the relevant paragraph:

> Supporting Pickle will enable wider usage in the Python ecosystem because Python's ecosystems of libraries depend strongly on the presence of protocols. Without these protocols, it's necessary for each library to implement a custom serialization method for every other library. For example, Dask Distributed has a custom serialization method for Keras at [distributed/protocol/keras.py]. See "[Pickle isn't slow, it's a protocol]" for more detail (notably, this post focuses on having an efficient Pickle implementation for PyTorch).

Let me know if I should change anything in that paragraph.

[1]:https://github.com/adriangb/community/blob/keras-pickle-edits/rfcs/20200902-pickle-for-keras.md
[distributed/protocol/keras.py]:https://blog.dask.org/2018/07/23/protocols-pickle
[Pickle isn't slow, it's a protocol]:https://blog.dask.org/2018/07/23/protocols-pickle

**Comment by mrocklin** [comment]
I'm happy to see this work.  I'm a bit slammed at the moment, but maybe instead of reviewing I could try to conscript @jakirkham ?  He has been thinking a lot about serialization recently, and I think that it would be good to have someone from the RAPIDS team engaged regardless.

**Comment by jakirkham** [comment]
Would suggest leveraging [pickle protocol 5]( https://docs.python.org/3/library/pickle.html#out-of-band-buffers ) where possible to avoid copying frames before sending them over the wire. This is something Dask ( https://github.com/dask/distributed/pull/3784 ) is able to leverage for zero-copy transmission.

**Comment by adriangb** [comment]
That is awesome, I didn't know that existed, thank you @jakirkham!

I think we'd need to implement protocol < 5 in any case because we'd need to support TF users with Python <3.8. So probably we can mention in the proposal that there is the possibility of zero-copy transmission, but not make that the main focus.

**Comment by jakirkham** [comment]
Well there is the [`pickle5` backport package]( https://github.com/pitrou/pickle5-backport ) that supports all the way back to Python 3.5. If Tensorflow made that a dependency for older Python versions, then that should help handle the long tail of support.

Edit: We support it as an optional dependency in Dask ( https://github.com/dask/distributed/pull/3849 ).

**Comment by adriangb** [comment]
Worth mentioning as well!

In general, I feel that focusing on a simple bug free implementation is primary relative to performance. TensorFlow is sort of hard to work with, ex: they still support Python 2.7. That said, I think having the ability to _eventually_ have a performant implementation with support down to Python 3.6 would be a good argument for starting to work on pickle support in the first place.

**Comment by jakirkham** [comment]
FWIW the normal way I do this is launder things through NumPy arrays (using `np.asarray`) since they support pickle protocol 5 (including with `pickle5` backport support if available). If pickle protocol 5 is not available, then NumPy arrays just pickle through an older protocol.

**Comment by adriangb** [comment]
Would you mind elaborating? Is this something like `return unpickle_fn, (np.asarray(bytes), )`

**Comment by jakirkham** [comment]
Yep, exactly.

**Comment by jakirkham** [comment]
Actually would tweak it slightly to `return unpickle_fn, (np.asarray(memoryview(bytes)), )`.

Reason being `numpy.asarray` handles `bytes` objects like an ASCII string, which is a little odd. Using a `memoryview` first helps NumPy pickup on the underlying type of `uint8`.

```python
In [1]: import numpy as np

In [2]: b = b"abc"

In [3]: np.asarray(b)
Out[3]: array(b'abc', dtype='|S3')

In [4]: np.asarray(memoryview(b))
Out[4]: array([97, 98, 99], dtype=uint8)
```

**Comment by adriangb** [comment]
Interesting. So just by doing that instead of `return unpickle_fn, (bytes), )` we would get pickle protocol 5 support/zero-copy transmission for Python >=3.8? That would be pretty cool. Certainly more feasible than having a check for the Python version and such.

**Comment by jakirkham** [comment]
Yep, NumPy also supports the `pickle5` backport package. So if that's available, it will work for Python 3.5+.

Exactly this is why we do that as well. NumPy is a common dependency. So most people have it as well.

**Comment by jakirkham** [comment]
cc @quasiben (for vis)

File: rfcs/20200902-pickle-for-keras.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+``` python
+# tensorflow/python/.../training.py
+from tf.keras.models import load_model
+
+class Model:
+    ...
+  def __reduce__(self, protocol):
+    self.save(f"ram://tmp/saving/{id(self)")
+    b = tf.io.gfile.read_folder(f"ram://tmp/saving/{id(self)")
```
Comment (stsievert) [inline rfcs/20200902-pickle-for-keras.md]: Typo:

``` diff
- b = tf.io.gfile.read_folder(f"ram://tmp/saving/{id(self)")
+ b = tf.io.gfile.read_folder(f"ram://tmp/saving/{id(self)}")
```

That's in the the line above too.

Comment (adriangb) [reply rfcs/20200902-pickle-for-keras.md]: Thank you. Resolved.

**Comment by adriangb** [comment]
Hi folks, let's make a push to get this wrapped up this weekend so we can submit the PR in the TF repo sometime next week and start to get some feedback on that side. Please let me know if you see anything that would be a blocker to this.

**Comment by stsievert** [comment]
Doesn't the implementation rely on NumPy's Pickle 5 support? So can't `__reduce_ex__` be replaced with `__reduce__`?

I've deleted the metric tests. It's a minimal implementation. I think it's obvious that the tests will pass with `NewMetric`.

File: rfcs/20200902-pickle-for-keras.md [hunk 2]
```diff
@@ -150,11 +150,9 @@ assert m1 == m2  # TODO: or some other check
 For `tf.keras.Model`, we can use `SaveModel` as the backend for `__reduce_ex__`:
 
 ``` python
-# tensorflow/python/.../training.py
 from tf.keras.models import load_model
 
-class Model:
-    ...
+class NewModel(Model):
```
Comment (adriangb) [inline rfcs/20200902-pickle-for-keras.md]: I'm curious as to why the name change? I was envisioning these code blocks as representing "here's a pseudocode of what this would look like if implemented in TF" and not necessarily "here's how users can create a picklable Model" which is the first thought that came to mind when I saw `NewModel(Model)`.

**Comment by adriangb** [comment]
> Doesn't the implementation rely on NumPy's Pickle 5 support? So can't `__reduce_ex__` be replaced with `__reduce__`?
> 
I was thinking that, but I wasn't 100% sure and I also had the thought that (1) this is a private ecosystem method, not something users will really be looking at so how nice the name looks doesn't really matter and (2) `__reduce_ex__` will allow more flexibility if anyone wants to use the pickle version within the method in the future.

> I've deleted the metric tests. It's a minimal implementation. I think it's obvious that the tests will pass with NewMetric.

👍

**Comment by stsievert** [comment]
> (1) this is a private ecosystem method, not something users will really be looking at so how nice the name looks doesn't really matter

I'm concerned with maintenance cost. Will some developer waste an hour looking into the difference between `__reduce__` and `__reduce_ex__` because the `protocol` argument isn't used?

> (2) `__reduce_ex__` will allow more flexibility if anyone wants to use the pickle version within the method in the future.

I don't think "more flexibility" is relevant; I think `__reduce__` is as flexible as `__reduce_ex__` because because the `protocol` argument isn't used in this implementation.

**Comment by adriangb** [comment]
I do see you point, but I think there's arguments either way. If you feel strongly about it and are 100% sure that we can use `__reduce__` and still get the full benefits of protocol 5 support because we offload to numpy, then by all means switch it back, I do not intend to hold this back for something that minor.

File: rfcs/20200902-pickle-for-keras.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+  def __reduce_ex__(self, protocol):
+    self.save(f"ram://tmp/saving/{id(self)")
+    b = tf.io.gfile.read_folder(f"ram://tmp/saving/{id(self)}")
+    return self._reconstruct_pickle, (np.asarray(memoryview(b)), )
+
+  @classmethod
+  def _reconstruct_pickle(cls, obj):
+    tf.io.gfile.write_folder(f"ram://tmp/saving/{id(obj)}", b)
+    return load_model(temp_ram_location)
```
Comment (stsievert) [inline rfcs/20200902-pickle-for-keras.md:166]: `temp_ram_location` and `b` aren't defined.

Comment (adriangb) [reply rfcs/20200902-pickle-for-keras.md:166]: Good catch, fixed.

**Comment by adriangb** [comment]
I think we are pretty much good to go.

@stsievert, do you want me to change `__reduce_ex__` back to `__reduce__` before submitting this?

**Comment by stsievert** [comment]
> change `__reduce_ex__` back to `__reduce__`

I think @jakirkham is the right person to ask.

I'm pretty sure zero-copy transmission remains possible with `__reduce__`, but am not absolutely certain. The documentation in [PEP 574] seems to indicate that zero-copy transmission remains possible with `__reduce__`:

> We see that several conditions are required for the above to work:
> 
> * `__reduce__` or `__reduce_ex__` must be able to return something that indicates a serializable no-copy buffer view.
> * ...

[PEP 574]:https://www.python.org/dev/peps/pep-0574/#example

**Comment by adriangb** [comment]
Thank you, will wait on their response and then send this out.

**Comment by jakirkham** [comment]
Yeah it shouldn't matter if we are just using NumPy to handle out-of-band pickling.

**Comment by adriangb** [comment]
Thank you all for your contributions. I'm going to merge this and move this to the TF repo.

**Comment by adriangb** [comment]
@stsievert can you sign the Google CLA to fix https://github.com/tensorflow/community/pull/286#issuecomment-696305649 if you want to keep authorship of your commits?

**Comment by stsievert** [comment]
I don't mind losing Git authorship. Feel free to remove it.
